draft-ietf-tls-esni.md

```diff
@@ -476,6 +473,37 @@ contain a "cached_info" extension {{!RFC7924}} with a CachedObject
 entry whose CachedInformationType is "cert", since this indication
 would divulge the true server name.
 
+## Recommended Padding Scheme {#padding}
+
+Future extensions could reveal sensitive information through their length. Consequently, 
+padding should be flexible and support arbitrary extension changes. 
+
+For each field in the inner ClientHello, clients need to determine how much to
+pad given the semantics of that field.  For example, if a client can propose
+various ALPN values, it could add padding to round up to the longest of those.
+
+The target padding length of most ClientHello extensions can be determined without server help.
+However, the "server_name" extension could benefit from server input (ECHOConfig.maximum_name_length). 
+Clients SHOULD compute the padding for this extension as follows:
+
+1. If ECHOConfig.maximum_name_length is longer than the actual server_name then 
+clients SHOULD add padding to make up that difference.
+2. Otherwise, if ECHOConfig.maximum_name_length is zero or less than the length 
+of the actual server_name then round the server_name up to a multiple of 32 octets 
+and randomly add another 32 octets 50% of the time and then include that amount
+of additional padding.
```

Comment:
Suggest: "add at least another 32 octets".

Reply:
Or "add one or more blocks of 32 octets" just to exclude behaviours that undo the effect of rounding the padding to a multiple of 32 octets.

Reply:
👍 to N >= 1 blocks. Clients can choose the distribution from which they want to sample N.

Reply:
On second thoughts, we should probably match what's done in encrypted DNS, because otherwise we could create a potential leak from correlating the DNS request length with the ECHO length.

Reply:
Don't think copying DNS is correct. RFC 8467 calls for queries to be padded to a multiple of 128 octets and could be that the server-name is not the only sensitive name. 

I'm not sure there's a point in recommending against zero additional blocks of 32 - so long as you don't do zero too often it'd be ok. If you look back to the name lengths seen in the data I have it may be that zero additional blocks 50% of the time is ok. 

But mainly, I'd say we may be better to wait 'till we have a few implementations and then see what makes best sense.

Reply:
> Don't think copying DNS is correct. RFC 8467 calls for queries to be padded to a multiple of 128 octets and could be that the server-name is not the only sensitive name.

Can you show that there is no information leakage if the two aren't in sync?

> I'm not sure there's a point in recommending against zero additional blocks of 32 - so long as you don't do zero too often it'd be ok. If you look back to the name lengths seen in the data I have it may be that zero additional blocks 50% of the time is ok.

The proposal above suggests N >= 1, not zero. In any case, as you suggest, spinning our wheels on something for which there is no backing data or analysis (yet) seems premature.

Reply:
Hmm, is this resolved?

Reply:
I actually don't think we should recommend the random padding. Let's do the minimum here.